File: res/nashorn2.js
var Nashorn2 = Java.type('com.winterbe.java8.samples.nashorn.Nashorn2');
var nashorn2 = new Nashorn2();
var result = nashorn2.fun('John Doe');
print('\n' + result);

nashorn2.fun2(123);
nashorn2.fun2(49.99);
nashorn2.fun2(true);
nashorn2.fun2("hi there")
nashorn2.fun2(String("bam"))
nashorn2.fun2(new Number(23));
nashorn2.fun2(new Date());
nashorn2.fun2(new RegExp());
nashorn2.fun2({foo: 'bar'});


print('passing object hash:');
nashorn2.fun3({
    k1: 'bar',
    k2: 'foo'
});


print('passing custom person object:');

function Person(firstName, lastName) {
    this.firstName = firstName;
    this.lastName = lastName;
    this.getFullName = function() {
        return this.firstName + " " + this.lastName;
    }
}

var person1 = new Person("Peter", "Parker");
nashorn2.fun3(person1);
nashorn2.fun4(person1);

File: src/com/winterbe/java8/samples/nashorn/Nashorn2.java
package com.winterbe.java8.samples.nashorn;

import jdk.nashorn.api.scripting.ScriptObjectMirror;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import java.io.FileReader;
import java.util.Arrays;

/**
 * Calling java methods from javascript with nashorn.
 *
 * @author Benjamin Winterberg
 */
public class Nashorn2 {

    public String fun(String name) {
        System.out.format("Hi there from Java, %s", name);
        return "greetings from java";
    }

    public void fun2(Object object) {
        System.out.println(object.getClass());
    }

    public void fun3(ScriptObjectMirror mirror) {
        System.out.println(mirror.getClassName() + ": " + Arrays.toString(mirror.getOwnKeys(true)));
    }

    public void fun4(ScriptObjectMirror person) {
        System.out.println("Full Name is: " + person.callMember("getFullName"));
    }

    public static void main(String[] args) throws Exception {
        ScriptEngine engine = new ScriptEngineManager().getEngineByName("nashorn");
        engine.eval(new FileReader("res/nashorn2.js"));
    }

}